In [23]:
import pandas as pd

train_df = pd.read_csv('Cargo Volume - Train.csv')
test_df = pd.read_csv('Cargo Volume - Test.csv')

# Data Inspection

In [24]:
train_df

Activity Period                 Operating Airline  \
0               201601              Sun Country Airlines   
1               200811                   United Airlines   
2               200709                    Qantas Airways   
3               201608               Philippine Airlines   
4               201708                Singapore Airlines   
...                ...                               ...   
27225           201108  United Airlines - Pre 07/01/2013   
27226           200703                   Asiana Airlines   
27227           201204                        US Airways   
27228           201005                        Air Canada   
27229           202002                  Korean Air Lines   

      Operating Airline IATA Code                 Published Airline  \
0                              SY              Sun Country Airlines   
1                              UA                   United Airlines   
2                              QF                    Qantas Airways   
3                              PR               Philippine Airlines   
4                              SQ                Singapore Airlines   
...                           ...                               ...   
27225                          UA  United Airlines - Pre 07/01/2013   
27226                          OZ                   Asiana Airlines   
27227                          US                        US Airways   
27228                          AC                        Air Canada   
27229                          KE                  Korean Air Lines   

      Published Airline IATA Code    GEO Summary           GEO Region  \
0                              SY       Domestic                   US   
1                              UA       Domestic                   US   
2                              QF  International  Australia / Oceania   
3                              PR  International                 Asia   
4                              SQ  International                 Asia   
...                           ...            ...                  ...   
27225                          UA  International               Canada   
27226                          OZ  International                 Asia   
27227                          US       Domestic                   US   
27228                          AC  International               Canada   
27229                          KE  International                 Asia   

      Activity Type Code Cargo Type Code Cargo Aircraft Type  \
0               Enplaned           Cargo           Passenger   
1               Deplaned           Cargo           Passenger   
2               Deplaned         Express           Passenger   
3               Deplaned            Mail           Passenger   
4               Deplaned         Express           Passenger   
...                  ...             ...                 ...   
27225           Enplaned         Express           Passenger   
27226           Deplaned            Mail           Freighter   
27227           Deplaned           Cargo           Passenger   
27228           Deplaned           Cargo           Passenger   
27229           Deplaned           Cargo           Passenger   

       Cargo Metric TONS  
0               8.871509  
1              53.946194  
2               0.043999  
3              19.023530  
4               6.951420  
...                  ...  
27225           5.687690  
27226           3.473669  
27227          18.564941  
27228           9.814090  
27229         525.214368  

[27230 rows x 11 columns]

In [25]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27230 entries, 0 to 27229
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Activity Period              27230 non-null  int64  
 1   Operating Airline            27230 non-null  object 
 2   Operating Airline IATA Code  27199 non-null  object 
 3   Published Airline            27230 non-null  object 
 4   Published Airline IATA Code  27199 non-null  object 
 5   GEO Summary                  27230 non-null  object 
 6   GEO Region                   27230 non-null  object 
 7   Activity Type Code           27230 non-null  object 
 8   Cargo Type Code              27230 non-null  object 
 9   Cargo Aircraft Type          27230 non-null  object 
 10  Cargo Metric TONS            27230 non-null  float64
dtypes: float64(1), int64(1), object(9)
memory usage: 2.3+ MB


In [26]:
train_df.describe()

Activity Period  Cargo Metric TONS
count     27230.000000       27230.000000
mean     201277.964378         216.669953
std         451.796120         432.731356
min      200507.000000           0.000454
25%      200904.000000           8.446826
50%      201303.000000          60.774235
75%      201703.000000         247.236041
max      202012.000000       10801.278330

In [27]:
train_df.isna().sum()

Activity Period                 0
Operating Airline               0
Operating Airline IATA Code    31
Published Airline               0
Published Airline IATA Code    31
GEO Summary                     0
GEO Region                      0
Activity Type Code              0
Cargo Type Code                 0
Cargo Aircraft Type             0
Cargo Metric TONS               0
dtype: int64

In [28]:
# Since the IATA code columns contain essentially the same information as the Operating/Published Airline columns, I will drop the IATA code columns
# because they contain null values and are otherwise redundant
train_df.drop(columns=['Operating Airline IATA Code', 'Published Airline IATA Code'], inplace=True)
train_df.isna().sum()

Activity Period        0
Operating Airline      0
Published Airline      0
GEO Summary            0
GEO Region             0
Activity Type Code     0
Cargo Type Code        0
Cargo Aircraft Type    0
Cargo Metric TONS      0
dtype: int64

In [29]:
train_df

Activity Period                 Operating Airline  \
0               201601              Sun Country Airlines   
1               200811                   United Airlines   
2               200709                    Qantas Airways   
3               201608               Philippine Airlines   
4               201708                Singapore Airlines   
...                ...                               ...   
27225           201108  United Airlines - Pre 07/01/2013   
27226           200703                   Asiana Airlines   
27227           201204                        US Airways   
27228           201005                        Air Canada   
27229           202002                  Korean Air Lines   

                      Published Airline    GEO Summary           GEO Region  \
0                  Sun Country Airlines       Domestic                   US   
1                       United Airlines       Domestic                   US   
2                        Qantas Airways  International  Australia / Oceania   
3                   Philippine Airlines  International                 Asia   
4                    Singapore Airlines  International                 Asia   
...                                 ...            ...                  ...   
27225  United Airlines - Pre 07/01/2013  International               Canada   
27226                   Asiana Airlines  International                 Asia   
27227                        US Airways       Domestic                   US   
27228                        Air Canada  International               Canada   
27229                  Korean Air Lines  International                 Asia   

      Activity Type Code Cargo Type Code Cargo Aircraft Type  \
0               Enplaned           Cargo           Passenger   
1               Deplaned           Cargo           Passenger   
2               Deplaned         Express           Passenger   
3               Deplaned            Mail           Passenger   
4               Deplaned         Express           Passenger   
...                  ...             ...                 ...   
27225           Enplaned         Express           Passenger   
27226           Deplaned            Mail           Freighter   
27227           Deplaned           Cargo           Passenger   
27228           Deplaned           Cargo           Passenger   
27229           Deplaned           Cargo           Passenger   

       Cargo Metric TONS  
0               8.871509  
1              53.946194  
2               0.043999  
3              19.023530  
4               6.951420  
...                  ...  
27225           5.687690  
27226           3.473669  
27227          18.564941  
27228           9.814090  
27229         525.214368  

[27230 rows x 9 columns]

In [30]:
test_df

Activity Period      Operating Airline Operating Airline IATA Code  \
0              200907        United Airlines                          UA   
1              202001  Nippon Cargo Airlines                          KZ   
2              201107         China Airlines                          CI   
3              201809         China Airlines                          CI   
4              200811   China Cargo Airlines                          CK   
...               ...                    ...                         ...   
6802           201603             Air Canada                          AC   
6803           201910        British Airways                          BA   
6804           201503        Air New Zealand                          NZ   
6805           201009          World Airways                          WO   
6806           200612        Delta Air Lines                          DL   

          Published Airline Published Airline IATA Code    GEO Summary  \
0           United Airlines                          UA       Domestic   
1     Nippon Cargo Airlines                          KZ  International   
2            China Airlines                          CI  International   
3            China Airlines                          CI  International   
4      China Cargo Airlines                          CK  International   
...                     ...                         ...            ...   
6802             Air Canada                          AC  International   
6803        British Airways                          BA  International   
6804        Air New Zealand                          NZ  International   
6805          World Airways                          WO       Domestic   
6806        Delta Air Lines                          DL       Domestic   

               GEO Region Activity Type Code Cargo Type Code  \
0                      US           Deplaned            Mail   
1                    Asia           Deplaned           Cargo   
2                    Asia           Enplaned           Cargo   
3                    Asia           Deplaned           Cargo   
4                    Asia           Enplaned           Cargo   
...                   ...                ...             ...   
6802               Canada           Enplaned            Mail   
6803               Europe           Enplaned           Cargo   
6804  Australia / Oceania           Deplaned            Mail   
6805                   US           Enplaned           Cargo   
6806                   US           Enplaned            Mail   

     Cargo Aircraft Type  
0              Passenger  
1              Freighter  
2              Passenger  
3              Passenger  
4              Freighter  
...                  ...  
6802           Passenger  
6803           Passenger  
6804           Passenger  
6805           Freighter  
6806           Passenger  

[6807 rows x 10 columns]

In [31]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6807 entries, 0 to 6806
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Activity Period              6807 non-null   int64 
 1   Operating Airline            6807 non-null   object
 2   Operating Airline IATA Code  6793 non-null   object
 3   Published Airline            6807 non-null   object
 4   Published Airline IATA Code  6793 non-null   object
 5   GEO Summary                  6807 non-null   object
 6   GEO Region                   6807 non-null   object
 7   Activity Type Code           6807 non-null   object
 8   Cargo Type Code              6807 non-null   object
 9   Cargo Aircraft Type          6807 non-null   object
dtypes: int64(1), object(9)
memory usage: 531.9+ KB


In [32]:
test_df.describe()

Activity Period
count      6807.000000
mean     201262.874247
std         452.272097
min      200507.000000
25%      200901.000000
50%      201212.000000
75%      201701.000000
max      202012.000000

In [33]:
test_df.isna().sum()

Activity Period                 0
Operating Airline               0
Operating Airline IATA Code    14
Published Airline               0
Published Airline IATA Code    14
GEO Summary                     0
GEO Region                      0
Activity Type Code              0
Cargo Type Code                 0
Cargo Aircraft Type             0
dtype: int64

In [34]:
# For the same reason as the training data, I will drop the IATA code columns from the test dataset
test_df.drop(columns=['Operating Airline IATA Code', 'Published Airline IATA Code'], inplace=True)
test_df.isna().sum()

Activity Period        0
Operating Airline      0
Published Airline      0
GEO Summary            0
GEO Region             0
Activity Type Code     0
Cargo Type Code        0
Cargo Aircraft Type    0
dtype: int64

In [35]:
test_df

Activity Period      Operating Airline      Published Airline  \
0              200907        United Airlines        United Airlines   
1              202001  Nippon Cargo Airlines  Nippon Cargo Airlines   
2              201107         China Airlines         China Airlines   
3              201809         China Airlines         China Airlines   
4              200811   China Cargo Airlines   China Cargo Airlines   
...               ...                    ...                    ...   
6802           201603             Air Canada             Air Canada   
6803           201910        British Airways        British Airways   
6804           201503        Air New Zealand        Air New Zealand   
6805           201009          World Airways          World Airways   
6806           200612        Delta Air Lines        Delta Air Lines   

        GEO Summary           GEO Region Activity Type Code Cargo Type Code  \
0          Domestic                   US           Deplaned            Mail   
1     International                 Asia           Deplaned           Cargo   
2     International                 Asia           Enplaned           Cargo   
3     International                 Asia           Deplaned           Cargo   
4     International                 Asia           Enplaned           Cargo   
...             ...                  ...                ...             ...   
6802  International               Canada           Enplaned            Mail   
6803  International               Europe           Enplaned           Cargo   
6804  International  Australia / Oceania           Deplaned            Mail   
6805       Domestic                   US           Enplaned           Cargo   
6806       Domestic                   US           Enplaned            Mail   

     Cargo Aircraft Type  
0              Passenger  
1              Freighter  
2              Passenger  
3              Passenger  
4              Freighter  
...                  ...  
6802           Passenger  
6803           Passenger  
6804           Passenger  
6805           Freighter  
6806           Passenger  

[6807 rows x 8 columns]

# Data Preprocessing
Much of the data is categorical. Many machine learning algorithms require numerical data. The data we are working with is nominal data, so using the one-hot encoding technique makes the most 
sense for the encoding strategy.

In [36]:
from sklearn.preprocessing import OneHotEncoder
label_encoder = OneHotEncoder(sparse_output=False)
train_df_cat_enc = label_encoder.fit_transform(train_df[['Operating Airline', 'Published Airline', 'GEO Summary', 'GEO Region', 'Activity Type Code', 'Cargo Type Code',
       'Cargo Aircraft Type']])
train_df_enc = pd.DataFrame(train_df_cat_enc)
train_df_enc['Activity Period'] = train_df['Activity Period']
train_df_enc['Cargo Metric TONS'] = train_df['Cargo Metric TONS']
train_df_enc

0    1    2    3    4    5    6    7    8    9  ...  192  193  194  \
0      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  1.0   
1      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  1.0   
2      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  0.0   
3      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  0.0   
4      0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  0.0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
27225  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
27226  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  0.0   
27227  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  1.0   
27228  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  1.0  0.0  1.0   
27229  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  1.0   

       195  196  197  198  199  Activity Period  Cargo Metric TONS  
0      0.0  0.0  0.0  0.0  1.0           201601           8.871509  
1      0.0  0.0  0.0  0.0  1.0           200811          53.946194  
2      1.0  0.0  0.0  0.0  1.0           200709           0.043999  
3      0.0  1.0  0.0  0.0  1.0           201608          19.023530  
4      1.0  0.0  0.0  0.0  1.0           201708           6.951420  
...    ...  ...  ...  ...  ...              ...                ...  
27225  1.0  0.0  0.0  0.0  1.0           201108           5.687690  
27226  0.0  1.0  0.0  1.0  0.0           200703           3.473669  
27227  0.0  0.0  0.0  0.0  1.0           201204          18.564941  
27228  0.0  0.0  0.0  0.0  1.0           201005           9.814090  
27229  0.0  0.0  0.0  0.0  1.0           202002         525.214368  

[27230 rows x 202 columns]

In [37]:
test_df_cat_enc = label_encoder.fit_transform(test_df[['Operating Airline', 'Published Airline', 'GEO Summary', 'GEO Region', 'Activity Type Code', 'Cargo Type Code',
       'Cargo Aircraft Type']])
test_df_enc = pd.DataFrame(test_df_cat_enc)
test_df_enc['Activity Period'] = test_df['Activity Period']
test_df_enc

0    1    2    3    4    5    6    7    8    9  ...  186  187  188  \
0     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  1.0  0.0   
1     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
2     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
3     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
4     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
6802  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
6803  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  1.0   
6804  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
6805  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  1.0   
6806  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  1.0  0.0  1.0   

      189  190  191  192  193  194  Activity Period  
0     0.0  0.0  1.0  0.0  0.0  1.0           200907  
1     1.0  0.0  0.0  0.0  1.0  0.0           202001  
2     1.0  0.0  0.0  0.0  0.0  1.0           201107  
3     1.0  0.0  0.0  0.0  0.0  1.0           201809  
4     1.0  0.0  0.0  0.0  1.0  0.0           200811  
...   ...  ...  ...  ...  ...  ...              ...  
6802  0.0  0.0  1.0  0.0  0.0  1.0           201603  
6803  1.0  0.0  0.0  0.0  0.0  1.0           201910  
6804  0.0  0.0  1.0  0.0  0.0  1.0           201503  
6805  1.0  0.0  0.0  0.0  1.0  0.0           201009  
6806  0.0  0.0  1.0  0.0  0.0  1.0           200612  

[6807 rows x 196 columns]

In [85]:
encoded_train_df = pd.get_dummies(train_df)
encoded_test_df = pd.get_dummies(test_df)
encoded_train_df, encoded_test_df = encoded_train_df.align(encoded_test_df, axis=1, join='outer', fill_value=0)

In [87]:
encoded_train_df

Activity Period  Activity Type Code_Deplaned  \
0               201601                            0   
1               200811                            1   
2               200709                            1   
3               201608                            1   
4               201708                            1   
...                ...                          ...   
27225           201108                            0   
27226           200703                            1   
27227           201204                            1   
27228           201005                            1   
27229           202002                            1   

       Activity Type Code_Enplaned  Cargo Aircraft Type_Combi  \
0                                1                          0   
1                                0                          0   
2                                0                          0   
3                                0                          0   
4                                0                          0   
...                            ...                        ...   
27225                            1                          0   
27226                            0                          0   
27227                            0                          0   
27228                            0                          0   
27229                            0                          0   

       Cargo Aircraft Type_Freighter  Cargo Aircraft Type_Passenger  \
0                                  0                              1   
1                                  0                              1   
2                                  0                              1   
3                                  0                              1   
4                                  0                              1   
...                              ...                            ...   
27225                              0                              1   
27226                              1                              0   
27227                              0                              1   
27228                              0                              1   
27229                              0                              1   

       Cargo Metric TONS  Cargo Type Code_Cargo  Cargo Type Code_Express  \
0               8.871509                      1                        0   
1              53.946194                      1                        0   
2               0.043999                      0                        1   
3              19.023530                      0                        0   
4               6.951420                      0                        1   
...                  ...                    ...                      ...   
27225           5.687690                      0                        1   
27226           3.473669                      0                        0   
27227          18.564941                      1                        0   
27228           9.814090                      1                        0   
27229         525.214368                      1                        0   

       Cargo Type Code_Mail  ...  Published Airline_Tradewinds Airlines  \
0                         0  ...                                      0   
1                         0  ...                                      0   
2                         0  ...                                      0   
3                         1  ...                                      0   
4                         0  ...                                      0   
...                     ...  ...                                    ...   
27225                     0  ...                                      0   
27226                     1  ...                                      0   
27227                     0  ...                                      0   
27228                     0  ...                 

In [86]:
encoded_test_df

Activity Period  Activity Type Code_Deplaned  \
0              200907                            1   
1              202001                            1   
2              201107                            0   
3              201809                            1   
4              200811                            0   
...               ...                          ...   
6802           201603                            0   
6803           201910                            0   
6804           201503                            1   
6805           201009                            0   
6806           200612                            0   

      Activity Type Code_Enplaned  Cargo Aircraft Type_Combi  \
0                               0                          0   
1                               0                          0   
2                               1                          0   
3                               0                          0   
4                               1                          0   
...                           ...                        ...   
6802                            1                          0   
6803                            1                          0   
6804                            0                          0   
6805                            1                          0   
6806                            1                          0   

      Cargo Aircraft Type_Freighter  Cargo Aircraft Type_Passenger  \
0                                 0                              1   
1                                 1                              0   
2                                 0                              1   
3                                 0                              1   
4                                 1                              0   
...                             ...                            ...   
6802                              0                              1   
6803                              0                              1   
6804                              0                              1   
6805                              1                              0   
6806                              0                              1   

      Cargo Metric TONS  Cargo Type Code_Cargo  Cargo Type Code_Express  \
0                     0                      0                        0   
1                     0                      1                        0   
2                     0                      1                        0   
3                     0                      1                        0   
4                     0                      1                        0   
...                 ...                    ...                      ...   
6802                  0                      0                        0   
6803                  0                      1                        0   
6804                  0                      0                        0   
6805                  0                      1                        0   
6806                  0                      0                        0   

      Cargo Type Code_Mail  ...  Published Airline_Tradewinds Airlines  \
0                        1  ...                                      0   
1                        0  ...                                      0   
2                        0  ...                                      0   
3                        0  ...                                      0   
4                        0  ...                                      0   
...                    ...  ...                                    ...   
6802                     1  ...                                      0   
6803                     0  ...                                      0   
6804                     1  ...                                      0   
6805                     0  ...                                      0   
6806                     1  ... 

We now need to split the test data further into a training set and a validation set. I will use and 80/20 split for training/test data.

In [88]:
import numpy as np

def split_train_test(df, test_ratio):
    shuffled_indices = np.random.permutation(len(df))
    test_set_size = int(len(df) * test_ratio)
    test_indices = shuffled_indices[:test_set_size]
    train_indices = shuffled_indices[test_set_size:]
    return df.iloc[train_indices], df.iloc[test_indices]

train_set, test_set = split_train_test(encoded_train_df, 0.2)

# Build And Evaluate Models
We are trying to predict the cargo weight of any given shipment. Given that we know the target, and the target is a real a number, there are a few algorithms that are best suited for this problem.
I will be evaluating the performance of a linear regression model, a Ridge model, a LASSO mdoel, a basic decision tree, and XG Boost.

In [89]:
train_y = train_set['Cargo Metric TONS']
test_y = test_set['Cargo Metric TONS']
train_x = train_set.drop(columns=['Cargo Metric TONS'])
test_x = test_set.drop(columns=['Cargo Metric TONS'])

In [90]:
train_y

24932     80.771191
22565     89.145101
15690      1.881986
20325     12.101594
9016       8.166161
            ...    
24128      0.483991
8707     147.900816
7098     125.397266
2585      54.552658
13500     24.435886
Name: Cargo Metric TONS, Length: 21784, dtype: float64

In [91]:
from sklearn.linear_model import LinearRegression

reg = LinearRegression()
reg.fit(train_x, train_y)

LinearRegression()

In [92]:
from sklearn.metrics import mean_squared_error
import numpy as np

hyp = reg.predict(test_x)
mse = mean_squared_error(test_y, hyp)
lin_reg_error = np.sqrt(mse)
lin_reg_error

294.61005101834064

In [93]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV

param = {'alpha': [0.01, 0.1, 1, 2, 3, 4, 10, 15, 20, 25, 100]}
ridge = Ridge()
best_ridge = GridSearchCV(ridge, param, scoring='neg_mean_squared_error', cv=5)
best_ridge.fit(train_x, train_y)

GridSearchCV(cv=5, estimator=Ridge(),
             param_grid={'alpha': [0.01, 0.1, 1, 2, 3, 4, 10, 15, 20, 25, 100]},
             scoring='neg_mean_squared_error')

In [94]:
mse = mean_squared_error(test_y, best_ridge.predict(test_x))
ridge_err = np.sqrt(mse)
ridge_err

294.7527196910736

In [95]:
from sklearn.linear_model import Lasso

lasso = Lasso()
best_lasso = GridSearchCV(lasso, param, scoring='neg_mean_squared_error', cv=5)
best_lasso.fit(train_x, train_y)

/opt/homebrew/lib/python3.11/site-packages/sklearn/linear_model/_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.263e+08, tolerance: 3.131e+05
  model = cd_fast.enet_coordinate_descent(
/opt/homebrew/lib/python3.11/site-packages/sklearn/linear_model/_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.853e+08, tolerance: 3.310e+05
  model = cd_fast.enet_coordinate_descent(
/opt/homebrew/lib/python3.11/site-packages/sklearn/linear_model/_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.375e+08, toleranc

GridSearchCV(cv=5, estimator=Lasso(),
             param_grid={'alpha': [0.01, 0.1, 1, 2, 3, 4, 10, 15, 20, 25, 100]},
             scoring='neg_mean_squared_error')

In [96]:
best_lasso_err = np.sqrt(mean_squared_error(test_y, best_lasso.predict(test_x)))
best_lasso_err

294.73297835149884

In [97]:
from sklearn.tree import DecisionTreeRegressor

clf = DecisionTreeRegressor()
clf.fit(train_x, train_y)

DecisionTreeRegressor()

In [98]:
mse = mean_squared_error(test_y, clf.predict(test_x))
dec_tree_err = np.sqrt(mse)
dec_tree_err

215.2214448463679

In [99]:
from xgboost import XGBRegressor

xgb = XGBRegressor()
xgb.fit(train_x, train_y)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             multi_strategy=None, n_estimators=None, n_jobs=None,
             num_parallel_tree=None, random_state=None, ...)

In [100]:
xgb_err = np.sqrt(mean_squared_error(test_y, xgb.predict(test_x)))
xgb_err

129.84121978542345

# Evaluate Models
The RMSE for each model is as follows:
1. XG Boost - 129.84121978542345
2. Decision Tree - 215.2214448463679
3. Linear Regression - 294.61005101834064
4. LASSO - 294.73297835149884
5. Ridge - 294.7527196910736

After evaluating all the models, we can see that the XG Boost model produced the best results. In order to generate our predictions on the test dataset, we will use this model. 


# Prediction

In [102]:
prediction = xgb.predict(encoded_test_df.drop(columns=['Cargo Metric TONS']))
prediction

array([123.90662 , 243.19292 , 122.87792 , ...,  21.169542, 496.39612 ,
        82.65026 ], dtype=float32)

In [105]:
pd.DataFrame(prediction, columns=['Cargo Metric TONS']).to_csv('prediction.csv', header='Cargo Metric TONS', index=False)